pkg/resourcesets/collector.go

```diff
@@ -248,97 +236,50 @@ func (h *ResourceHandler) filterByLabel(ctx context.Context, dr dynamic.Resource
+		includeRegex = regexp.MustCompile(filter.ResourceNameRegexp)
 	}
-
-	// filter out using ExcludeResourceNameRegexp
+	var excludeRegex *regexp.Regexp
 	if filter.ExcludeResourceNameRegexp != "" {
-		if filter.ResourceNameRegexp == "" {
-			filteredByName = resourceObjectsList
-		}
-		var newFilteredByName []unstructured.Unstructured
-		for _, resObj := range filteredByName {
-			metadata := resObj.Object["metadata"].(map[string]interface{})
-			name := metadata["name"].(string)
-			nameMatched, err := regexp.MatchString(filter.ExcludeResourceNameRegexp, name)
-			if err != nil {
-				return filteredByName, err
-			}
-			if nameMatched {
-				logrus.Debugf("Skipping [%s] because it did match ExcludeResourceNameRegexp [%s]", name, filter.ExcludeResourceNameRegexp)
-				key := resObj.Object["filterUid"].(string)
-				filteredByNameMap[key] = false
-				continue
-			}
-			newFilteredByName = append(newFilteredByName, resObj)
-			key := resObj.Object["filterUid"].(string)
-			filteredByNameMap[key] = true
-		}
-		filteredByName = newFilteredByName
-	}
-	// filter by names as fieldSelector:
-	if len(filter.ResourceNames) > 0 {
-		logrus.Debugf("Using ResourceNames [%s] to filter resource names", strings.Join(filter.ResourceNames, ","))
-		allowedNames := make(map[string]bool)
-		for _, name := range filter.ResourceNames {
-			allowedNames[name] = true
-		}
-		for _, resObj := range resourceObjectsList {
-			metadata := resObj.Object["metadata"].(map[string]interface{})
-			name := metadata["name"].(string)
-			if allowedNames[name] {
-				filteredByResourceNames = append(filteredByResourceNames, resObj)
-			}
-		}
+		excludeRegex = regexp.MustCompile(filter.ExcludeResourceNameRegexp)
+	}
 
-		if len(filteredByResourceNames) == 0 {
-			// exact names were provided, but no resources found by that name
-			// so return anything obtained from matching resource names by regex
-			// if that list is empty too, means nothing matched the filters
-			return filteredByName, nil
-		}
-		if len(filteredByNameMap) > 0 {
-			// avoid duplicates
-			for _, resObj := range filteredByResourceNames {
-				key := resObj.Object["filterUid"].(string)
-				if !filteredByNameMap[key] {
-					filteredByName = append(filteredByName, resObj)
+	for resObjID := range *resourceObjectsList {
+		// Check if the name is in the includeNames list
+		name := (*resourceObjectsList)[resObjID].GetName()
+		if includeNameMap[name] {
+			filteredByName = append(filteredByName, (*resourceObjectsList)[resObjID])
+			logrus.Debugf("Including [%s] because it mathed a name in ResourceNames", name)
+		} else {
+			if filter.ResourceNameRegexp != "" || filter.ExcludeResourceNameRegexp != "" {
+				// We expect this to be true, when set and match or when nil (true by default)
+				includeMatch := includeRegex == nil || includeRegex.MatchString(name)
+				// We expect this to be true, only when regex match (false by default)
+				excludeMatch := excludeRegex != nil && excludeRegex.MatchString(name)
```

Comment:
You can move the `excludeMatch` check inside
```
if includeMatch {
  excludeMatch := excludeRegex != nil && excludeRegex.MatchString(name)
  if !excludeMatch {
     filteredByName = append(...)
     logrs.Debugf("Including...")
 }
 ```
 
 This saves doing a pattern-match when it will never be used.